lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/TxWithUTxO.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE ExplicitNamespaces #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- |
+-- Copyright: © 2024 Cardano Foundation
+-- License: Apache-2.0
+--
+-- Provides the 'TxWithUTxO' data type.
+--
+module Internal.Cardano.Write.Tx.TxWithUTxO
+    ( type TxWithUTxO
+    , pattern TxWithUTxO
+    , isValid
+    )
+    where
+
+import Prelude
+
+import Cardano.Ledger.Api
+    ( AlonzoEraTxBody (collateralInputsTxBodyL)
+    , EraTx (bodyTxL)
+    , EraTxBody (TxBody, inputsTxBodyL)
+    )
+import Control.Lens
+    ( over
+    , view
+    )
+import Data.Set
+    ( Set
+    )
+import Internal.Cardano.Write.Tx
+    ( IsRecentEra
+    , Tx
+    , TxIn
+    , UTxO (UTxO)
+    )
+
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+
+-- | A transaction with an associated UTxO set.
+--
+-- Every input in the transaction is guaranteed to resolve to a UTxO within the
+-- associated UTxO set.
+--
+-- The UTxO set may also contain additional UTxOs that are not referenced by
+-- the transaction.
+--
+data TxWithUTxO era = UnsafeTxWithUTxO (Tx era) (UTxO era)
+
+{-# COMPLETE TxWithUTxO #-}
+pattern TxWithUTxO :: IsRecentEra era => Tx era -> UTxO era -> TxWithUTxO era
+pattern
+    TxWithUTxO tx utxo <- UnsafeTxWithUTxO tx utxo where
+    TxWithUTxO tx utxo = construct tx utxo
+
+-- | Constructs a 'TxWithUTxO' object.
+--
+-- The given transaction will be stripped of any inputs that do not resolve to
+-- UTxOs with the given UTxO set.
```

Comment:
> Why not fail instead of silently stripping?

That's a fair point.

I think the ability to strip unresolved inputs is useful for building shrinkers, so we do want to have that ability somewhere.

On reflection though, I don't think this should be the default behaviour.

I've made the following changes:
1. renamed `construct` to `constructFilter`;
1. added a second constructor called `constructEither` that will fail if there are any unresolved inputs;
1. removed the constructor pattern synonym.

Reply:
`constructFilter` — even with that name — seems very dangerous to me. Are there any drawbacks to moving it to the `Gen` module? Being required to `fromJust` the result of `constructFilter` instead of having access to `UnsafeTxWithUTxO` seems like a good thing to me if we did that. If we'd forget to update `removeUnresolvableInputs`, then `unresolvableInputs`'s `allInputsTxBodyF` would catch it and cause a failure.

We could also name it more in the style of `accursedUnutterablePerformIO` — e.g. `construct` and `constructBySilentlyDroppingUnresolvableInputs`... But moving it would seem safer / saner to me.

(I guess `TxWithUTxO` isn't a public module... but then again, if we want to expose it, creaing a public version containing everything but `constructFilter` seems very overkill to me)

Reply:
I've renamed these to:

```hs
construct         :: Tx era -> UTxO era -> Either (NESet TxIn) (TxWithUTxO era)
constructFiltered :: Tx era -> UTxO era ->                      TxWithUTxO era
```

- The `construct` function only succeeds if all inputs are resolveable. Otherwise, it returns a set of unresolvable inputs.
- The `constructFiltered` function always succeeds by filtering out any inputs from the transaction that are not resolvable.